src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <vector>
+#include <variant>
+
+namespace raksha::ir::auth_logic {
+
+class AstPrincipal {
+  public:
+    AstPrincipal(std::string name) : name_(std::move(name)) {}
+  private:
+    std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum AstSign { negated, positive };
+
+class AstPredicate {
+  public:
+    AstPredicate(
+        std::string name,
+        std::vector<std::string> args,
+        AstSign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        sign_(std::move(sign)) {}
```

Comment:
No need to `std::move` `sign`; it is an `enum` and will therefore be efficiently copiable.

Reply:
I might leave this around for a moment longer because it happens to prevent an unused private field `-Werr` about `sign_` but I left a TODO

Reply:
Why would `move` suppress the unused field error?

Reply:
I have no idea why it suppresses that error because this doesn't look like a use of `sign_` at all, but it does suppress the error!